### PR TITLE
BASISReduction: retain only events within a time window

### LINKED ISSUE #25448: BASISReduction: retain only events within a time window
users are not able to exclude retain events from only a particular time window

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**
- [x] Property to retain events only within a time window
- [x] Updated algorithm documentation
- [x] updated [release notes](https://github.com/mantidproject/mantid/blob/da52e7fed37ef2fc5e7ca4d18f8342e8f41bcb12/docs/source/release/v4.1.0/indirect_geometry.rst#id9)
- [x] test for reviewer

**Report to:** User: James Torres

**To test:**
Run the following script to output the proton charge after retaining events only from different time windows:
```
for time_window in ('0-3600', '3600-7200', '7200-end'):
    BASISReduction(RunNumbers='90177',
                   RetainTimeSegment='90177:'+time_window,
                   DoFluxNormalization=False)
    print(mtd['BSS_90177_sqw'].getRun().getProtonCharge())
```
Answers:

| time_window  | proton charge  (up to third decimal)|
|---|---|
| 0-3600  | 1383.709  |
| 3600-7200  | 1384.994  |
| 7200-end  | 5566.604  |

Fixes #25448

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/BASISReduction.py [hunk 1]
```diff
@@ -454,41 +456,66 @@ def _sum_runs(self, run_set, sam_ws, extra_ext=None):
...
+                      OutputWorkspace=name)
+        if str(run) + ':' in self.getProperty('RetainTimeSegment').value:
+            kwargs.update(self._retainEvents(run))
+        sapi.LoadEventNexus(**kwargs)
+        if str(run) + ':' in self.getProperty('ExcludeTimeSegment').value:
+            self._filterEvents(run, name)
+
     def _sum_flux(self, run_set, mon_ws, extra_ext=None):
-        self._aggregate_flux = 0.0
+        r"""
```
Comment (TTitcombe) [inline Framework/PythonInterface/plugins/algorithms/BASISReduction.py]: Is there any reason this doc string is a string literal and the one above isn't?
As far as I'm aware it's only necessary if the string contains a backslash as Python3 often interprets this as an invalid escape sequence

**Review by TTitcombe — COMMENTED** [review]
I'm unable to test as not at SNS (running the script complains about `/SNS/BSS/shared/autoreduce/new_masks_08_12_2015/BASIS_Mask_default_111.xml` not being found. Can I be sent the data necessary to run this script from ISIS?

The code side of things looks good. I have a minor comment about doc-strings but I don't think it's important enough to change

**Comment by martyngigg** [comment]
I think if this is to fix the BASIS tests then it needs to be able run the system tests outside of SNS.

File: Framework/PythonInterface/plugins/algorithms/BASISReduction.py (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/plugins/algorithms/BASISReduction.py]: @TTitcombe No other than "historical" reasons. I have changed all docstrings to Numpy style, and using string literals

**Comment by jmborr** [comment]
> I think if this is to fix the BASIS tests then it needs to be able run the system tests outside of SNS.

@martyngigg Indeed the system tests do not require access to SNS. It is the manual test for the reviewer of this PR which requires access to SNS.

**Comment by jmborr** [comment]
@TTitcombe @martyngigg I will recruit someone from SNS to run the manual test.

**Comment by AndreiSavici** [comment]
I've tried the manual test. The results look OK

**Comment by martyngigg** [comment]
Thanks @jmborr!

**Pull request merged**